### PR TITLE
Baseline framing

### PR DESCRIPTION
**Pull request by jorshi**
Builds on top of the basic mel spectrogram baseline and implements framing and batching.

### CODE REVIEW

File: baseline/baseline.py [hunk 1]
```diff
@@ -6,54 +6,76 @@
...
-        newmin = torch.iinfo(torch.int8).min
-        x20 = torch.tensor(
-            (x20 * (newmax - newmin) + newmin), dtype=torch.int8, device=x20.device
-        )
+
+        # Scale to int8 value range and cast to int
+        int8_max = torch.iinfo(torch.int8).max
+        int8_min = torch.iinfo(torch.int8).min
+        x20 = x20 * (int8_max - int8_min) + int8_min
+        x20 = x20.type(torch.int8)
```
Comment (jorshi) [inline baseline/baseline.py:77]: I was testing this with a some different input signals and I don't think this is working quite how we expect. The ranges of the embedded signals after projection seem to be getting pretty extreme (each projection seems to emphasize that difference even more), even though it looks like the projection matrices are being normalized. So by the time we get to x20 the values are pretty only -128 or 127, we are capturing the range in that embedding very well.

File: baseline/baseline.py [hunk 2]
```diff
@@ -82,79 +105,169 @@ def load_model(model_file_path: str, device: str = "cpu") -> Any:
...
+    """
+    batch_size, num_samples = audio.shape
+
+    # Adjust the number of samples if centering to allow for a half
+    # frame size number of padded samples padded at the start.
+    start_pad = 0
+    if is_centered:
+        half_frame_size = int(frame_size // 2)
+        num_samples += half_frame_size
+        start_pad = half_frame_size
```
Comment (jorshi) [inline baseline/baseline.py:130]: @maxsolomonhenry -- modified this a little bit, we were overpadding non-centered signals so I tried this approach of adjusting the num_samples by half_frame_size if centering prior to calculating the number of samples to append to the end.

Our implementation right now isn't quite in alignment with how librosa handles framing. Since we are padding the end to make sure that the entire frame from the last hop is entirely contained, we are ending up padding more at the end then I think we need to. When looking at the corresponding timestamps we end up with centers of frames that extend beyond the duration of the input. 

The way that librosa handles this is to not zero pad at all when not centering. In this case the input can be truncated a bit. When centering they pad both the start and end by half a frame and then applying framing (again no zero padding occurs during framing).

What we have here works fine, but it is a bit confusing to get a list of timestamps back that extends past the duration of the input.

Comment (turian) [reply baseline/baseline.py:130]: @jorshi Can you we use torchaudio to do this for us?

File: baseline/baseline.py [hunk 3]
```diff
@@ -82,79 +105,169 @@ def load_model(model_file_path: str, device: str = "cpu") -> Any:
...
+            result = model(batch[0])
+            for size, embedding in result.items():
+                embeddings[size].append(embedding)
+
+    # Concatenate mini-batches back together and unflatten the frames back
+    # to audio batches
+    embeddings = dict(embeddings)
+    for size, embedding in embeddings.items():
+        embeddings[size] = torch.cat(embedding, dim=0)
+        embeddings[size] = embeddings[size].unflatten(0, (audio_batches, num_frames))
```
Comment (turian) [inline baseline/baseline.py:263]: It's still very hard for me to tell if this is unframing the audio correctly for each sample. I suspect it's not correct but it's hard to tell just by inspection. I think we need some sort of test :\

File: baseline/baseline.py (hunk 2 above)
Comment (jorshi) [reply baseline/baseline.py:130]: @turian doesn't look like there are any utils for framing in torchaudio

Comment (maxsolomonhenry) [reply baseline/baseline.py:130]: @jorshi If we want a centered frame at t = 0 we need to pad up front. As for the end, if you don't want time stamps that extend beyond the signal we'll just have to truncate the output by one frame.

```
if end_pad > half_frame_size:
    # Truncate `audio` instead of padding it.

File: baseline/baseline.py [hunk 4]
```diff
@@ -82,79 +105,169 @@ def load_model(model_file_path: str, device: str = "cpu") -> Any:
...
+        center: whether of not the frames were padded to center audio, if false
+            an offset will be applied equal to frame_size // 2.
+
+    Returns:
+        Timestamps in seconds.
+    """
+
+    offset = 0
+    if not center:
+        if frame_size is None:
```
Comment (maxsolomonhenry) [inline baseline/baseline.py:173]: I don't think there's a need for this. Timestamps will be the same whether or not the frames are centered.

File: baseline/baseline.py [hunk 5]
```diff
@@ -82,79 +105,169 @@ def load_model(model_file_path: str, device: str = "cpu") -> Any:
...
-    dataset = torch.utils.data.TensorDataset(audio)
+    # Split the input audio signals into frames and then flatten to create a tensor
+    # of audio frames that can be batch processed. We will unflatten back out to
+    # (audio_baches, num_frames, embedding_size) after creating embeddings.
+    frames = frame_audio(audio, model.n_fft, hop_size, center)
+    audio_batches, num_frames, frame_size = frames.shape
+    frames = frames.flatten(end_dim=1)
+
+    # We're using a DataLoader to help with batching of frames
+    dataset = torch.utils.data.TensorDataset(frames)
```
Comment (maxsolomonhenry) [inline baseline/baseline.py:243]: Dope.

File: baseline/baseline.py (hunk 3 above)
Comment (maxsolomonhenry) [reply baseline/baseline.py:263]: @turian What do you mean by "unframing the audio"?

File: baseline/baseline.py [hunk 6]
```diff
@@ -82,79 +105,169 @@ def load_model(model_file_path: str, device: str = "cpu") -> Any:
...
+    # Concatenate mini-batches back together and unflatten the frames back
+    # to audio batches
+    embeddings = dict(embeddings)
+    for size, embedding in embeddings.items():
+        embeddings[size] = torch.cat(embedding, dim=0)
+        embeddings[size] = embeddings[size].unflatten(0, (audio_batches, num_frames))
+
+    # Get timestamps in seconds.
+    timestamps = get_timestamps_for_embedding(
+        model.sample_rate, num_frames, hop_size, frame_size, center
```
Comment (maxsolomonhenry) [inline baseline/baseline.py:267]: Again, `center` is not required for this calculation.

File: baseline/baseline.py (hunk 1 above)
Comment (maxsolomonhenry) [reply baseline/baseline.py:77]: I don't know if this is helpful, but you might just do individual projections from the largest embedding size down to each smaller one, instead of cascading projections. I suggested this a little while ago.

Comment (jorshi) [reply baseline/baseline.py:77]: Ya good idea -- i'll try this out

File: baseline/baseline.py (hunk 4 above)
Comment (jorshi) [reply baseline/baseline.py:173]: I thought that we always calculated the timestamp for a frame as being the center of the frame -- at least for spectral stuff? Was referring to https://librosa.org/doc/latest/generated/librosa.frames_to_samples.html#librosa.frames_to_samples -- (which is used to calculate timestamps for frames).

Comment (jorshi) [reply baseline/baseline.py:173]: So if there isn't centering / zero-padding then the first timestamp will be frame_size // 2
